plugins/statistic.koplugin/main.lua

```diff
@@ -0,0 +1,423 @@
+    for curr_file in lfs.dir(statistic_dir) do
+        local path = statistic_dir .. "/" .. curr_file
+        if lfs.attributes(path, "mode") == "file" then
+            local book_result = self:importFromFile(curr_file)
+            if book_result and book_result.title ~= self.data.title then
+                table.insert(total_stats, {
+                    text = _(book_result.title),
+                    mandatory = os.date("!%X", tonumber(book_result.total_time)),
+                    callback = function()
+                        self.total_status:swithItemTable(nil, self:getDatesForBook(book_result))
+                        UIManager:show(self.total_menu)
+                        return true
+                    end,
+                })
+                total_books_time = total_books_time + book_result.total_time
+            end
+        end
+    end
+    total_books_time = total_books_time + tonumber(self.data.total_time)
+    table.insert(total_stats, 1, { text = _("All time"), mandatory = os.date("!%X", total_books_time) })
+    table.insert(total_stats, 2, { text = _("----------------------------------------------------") })
+    table.insert(total_stats, 3, {
+        text = _(self.data.title),
+        mandatory = os.date("!%X", tonumber(self.data.total_time)),
+        callback = function()
+            self.total_status:swithItemTable(nil, self:getDatesForBook(self.data))
+            UIManager:show(self.total_menu)
+            return true
+        end,
+    })
+    return total_stats
+end
+
+function ReaderStatistic:getBookProperties()
+    local props = self.view.document:getProps()
+    if props.title == "No document" or props.title == "" then --sometime crengine returns "No document" try to get one more time
+    props = self.view.document:getProps()
+    end
+    return props
+end
+
+function ReaderStatistic:onPageUpdate(pageno)
+    if self.is_enabled then
+        local curr_time = TimeVal:now()
+        local diff_time = curr_time.sec - self.last_time.sec
+
+        -- if last update was more then 10 minutes then current period set to 0
+        if (diff_time > 600) then
+            self.current_period = 0
+        end
+
+        if diff_time >= self.page_min_read_sec and diff_time <= self.page_max_read_sec then
+            self.current_period = self.current_period + diff_time
+            self.data.total_time = self.data.total_time + diff_time
+            local timeData = {
+                time = curr_time.sec,
+                read = diff_time,
+                page = pageno
+            }
+            table.insert(self.data.details, timeData)
+        end
+
+        self.last_time = curr_time
+    end
+end
+
+function ReaderStatistic:exportToFile(book_properties)
+    if book_properties then
+        self:savePropertiesInToData(book_properties)
+    end
+
+    local statistics = io.open(statistic_dir .. "/" .. self.data.title .. ".stat", "w")
+    if statistics then
+        local current_locale = os.setlocale()
```

Comment:
I get the intent here, I just hope this won't be an issue on slightly broken embedded systems ;).
Not sure you have a better way to deal with this though, so this is just just a random comment ;).